Add service and deployment create sub-commands to kubectl

asalkeld opened the pull request:
``` release-note
Added new commands kubectl create service & kubectl create deployment
```

Partial #25382

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/27109)

<!-- Reviewable:end -->

Commit:
Fix namespace_test error message

kargakis commented:
@kubernetes/kubectl

deads2k commented on pkg/kubectl/cmd/create_service.go:
```diff
@@ -0,0 +1,214 @@
+// NewCmdCreateServiceClusterIP is a command to create generic secrets from files, directories, or literal values
+func NewCmdCreateServiceClusterIP(f *cmdutil.Factory, cmdOut io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "clusterip NAME [--tcp=<port>:<targetPort>] [--dry-run]",
```
This is a neat idea.  Don't prompt for selector, auto-generate one based on name, then presumably you expect a `kubectl set selector` to exist?  I'm fine with it to start, but a description of your vision would be good.

asalkeld replied:
@deads2k happy to do that, but was probably assuming we would have --edit

deads2k commented:
I think its work considering a `--headless` flag for `kubectl create service clusterip` since the field is immutable.  @smarterclayton how common do you think it is?

asalkeld commented on pkg/kubectl/cmd/create_service.go:
```diff
@@ -0,0 +1,214 @@
+	cmdutil.AddPrinterFlags(cmd)
+	cmdutil.AddGeneratorFlags(cmd, cmdutil.ServiceClusterIPGeneratorV1Name)
+	addPortFlags(cmd)
+	cmd.Flags().Bool("headless", false, "Make this service 'headless' (do not do loadbalancing).")
```
@deads2k here is the headless flag, I'll add it to the example text to make it more obvious.

smarterclayton replied:
I'm really hesitant to enshrine "headless" in an argument - our use up to
this point has entirely been colloquial and not connected to the API, and
no other use refers to it as that.  Since you can only set cluster IP on
creation, a string flag "--cluster-ip=[<IP>|None]" with a good description
would be better than inventing a new term.  The description of the command
can describe what --cluster-ip=None does.

On Sun, Jun 26, 2016 at 4:36 AM, Angus Salkeld <notifications@github.com>
wrote:

> In pkg/kubectl/cmd/create_service.go
> <https://github.com/kubernetes/kubernetes/pull/27109#discussion_r68499495>
> :
>
> > +	cmd := &cobra.Command{
> > +		Use:     "clusterip NAME [--tcp=<port>:<targetPort>] [--dry-run]",
> > +		Short:   "Create a clusterIP service.",
> > +		Long:    serviceClusterIPLong,
> > +		Example: serviceClusterIPExample,
> > +		Run: func(cmd *cobra.Command, args []string) {
> > +			err := CreateServiceClusterIP(f, cmdOut, cmd, args)
> > +			cmdutil.CheckErr(err)
> > +		},
> > +	}
> > +	cmdutil.AddApplyAnnotationFlags(cmd)
> > +	cmdutil.AddValidateFlags(cmd)
> > +	cmdutil.AddPrinterFlags(cmd)
> > +	cmdutil.AddGeneratorFlags(cmd, cmdutil.ServiceClusterIPGeneratorV1Name)
> > +	addPortFlags(cmd)
> > +	cmd.Flags().Bool("headless", false, "Make this service 'headless' (do not do loadbalancing).")
>
> @deads2k <https://github.com/deads2k> here is the headless flag, I'll add
> it to the example text to make it more obvious.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27109/files/a3d7f495b814f2dd5a595ea1a8f4879ebfd9828a#r68499495>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p5S87x19y1PpV2IXty8zV1pql4zfks5qPjoVgaJpZM4Ixl7H>
> .
>

Commit:
Add "create service" sub-commands

This adds:
 create service clusterip
 create service nodeport
 create service loadbalancer

asalkeld commented:
@smarterclayton I have changed headless to clusterip=None, yikes looks like a bit of a clash with PR #19509

smarterclayton commented on pkg/kubectl/cmd/create_service.go:
```diff
@@ -0,0 +1,215 @@
+  kubectl create service clusterip my-cs --tcp=5678:8080
+
+  # Create a new clusterIP service named my-cs (in headless mode)
+  kubectl create service clusterip my-cs --tcp=5678:8080 --clusterip="None"`
```
Headless services can't be port mapped- might want to not show that in the example.

asalkeld commented:
Thanks @smarterclayton sorted that out, can you ptal? It would be great to get this in.

asalkeld commented:
Had to rebase as the Fake class now returns more fields

deads2k commented:
@fabianofranz ptal

smarterclayton commented:
The general concept and structure is ok to me - assigning to Fabiano for review

fabianofranz commented on pkg/kubectl/cmd/create_deployment.go:
```diff
@@ -0,0 +1,77 @@
+Create a deployment with the specified name.`
+
+	deploymentExample = `  # Create a new deployment named my-dep
+  kubectl create deployment my-dep --image=busybox`
```
Looks like there are some unwanted spaces on the left. You may want to use `dedent.Detent` like [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/label.go#L48).

fabianofranz commented on pkg/kubectl/cmd/create_deployment.go:
```diff
@@ -0,0 +1,77 @@
+	deploymentLong = `
+Create a deployment with the specified name.`
+
+	deploymentExample = `  # Create a new deployment named my-dep
```
"Create a new deployment named my-dep that runs the busybox image" or something like that.

fabianofranz commented on pkg/kubectl/cmd/create_deployment.go:
```diff
@@ -0,0 +1,77 @@
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddPrinterFlags(cmd)
+	cmdutil.AddGeneratorFlags(cmd, cmdutil.DeploymentBasicV1Beta1GeneratorName)
+	cmd.Flags().StringSlice("image", []string{}, "Image name to run.")
```
This flag is always required, right? Do `cmd.MarkFlagRequired` for completions.

asalkeld replied:
I'll add the flag.

fabianofranz commented on pkg/kubectl/cmd/create_deployment.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+// CreateDeployment implements the behavior to run the create deployment command
+func CreateDeployment(f *cmdutil.Factory, cmdOut io.Writer, cmd *cobra.Command, args []string) error {
```
If `--image` is always required, you need to validate that. [Here's an example](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/create_secret.go#L238-L243).

asalkeld replied:
OK, I did it a different way - not stressed which one is preferred.

fabianofranz replied:
With `ParamNames` works for me, tks.

fabianofranz replied:
Even though you validate with generator params, I think you still need `cmd.MarkFlagRequired` for command line completions. [Example](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/create_secret.go#L157).

fabianofranz commented on pkg/kubectl/cmd/create_deployment_test.go:
```diff
@@ -0,0 +1,39 @@
+	"testing"
+)
+
+func TestCreateDeployment(t *testing.T) {
```
Add a test where `--image` is not provided.

fabianofranz commented on pkg/kubectl/cmd/create_service.go:
```diff
@@ -0,0 +1,215 @@
+	serviceNodePortLong = `
+Create a nodeport service with the specified name.`
+
+	serviceNodePortExample = `  # Create a new nodeport service named my-ns
```
Indentation, like in my other comment.

fabianofranz replied:
Still needs `dedent.Dedent`.

fabianofranz commented on pkg/kubectl/cmd/create_service.go:
```diff
@@ -0,0 +1,215 @@
+  kubectl create service clusterip my-cs --clusterip="None"`
+)
+
+func addPortFlags(cmd *cobra.Command) {
```
In `kubect expose` (which creates a new service) we use two flags, `--port` and `--target-port`. Any reason not to use the same here?

asalkeld replied:
Because ports is a list, and this then will only allow one set which seems limiting. If we do it this way you can pass in multiple port mappings.

fabianofranz replied:
> Because ports is a list, and this then will only allow one set which seems limiting. 
> If we do it this way you can pass in multiple port mappings.

Makes sense, works for me.

asalkeld commented on pkg/kubectl/deployment.go:
```diff
@@ -0,0 +1,107 @@
+func (DeploymentBasicGeneratorV1) ParamNames() []GeneratorParam {
+	return []GeneratorParam{
+		{"name", true},
+		{"image", true},
```
This ^ "true" makes the param required. see: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/generate.go#L60-L71

asalkeld commented on pkg/kubectl/deployment_test.go:
```diff
@@ -0,0 +1,134 @@
+		{
+			params: map[string]interface{}{
+				"name":  "foo",
+				"image": []string{},
```
here is the test for image not given.

asalkeld commented:
@fabianofranz I have:
* fixed the indentation
* added a test to prove that image is required

Let me know if you still want the other way of validating required flags. I was just trying to
do it in a way that seemed to be consistent with the other generator commands.

fabianofranz commented on pkg/kubectl/cmd/create_service.go:
```diff
@@ -0,0 +1,217 @@
+	serviceLoadBalancerLong = `
+Create a LoadBalancer service with the specified name.`
+
+	serviceLoadBalancerExample = `  # Create a new nodeport service named my-lbs
```
Still needs `dedent.Dedent`.

asalkeld replied:
Yip, I'll add those and re-submit.

fabianofranz commented on pkg/kubectl/cmd/create_service.go:
```diff
@@ -0,0 +1,217 @@
+// NewCmdCreateServiceLoadBalancer is a macro command for creating secrets to work with Docker registries
+func NewCmdCreateServiceLoadBalancer(f *cmdutil.Factory, cmdOut io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "loadbalancer NAME [--tcp=port:targetPort] [--ingressIP=ipaddress] [--dry-run]",
```
Where are you declaring/using `ingressIP`? Also use either `--ingress-ip` or `--ingressip`.

asalkeld replied:
@fabianofranz thanks for the detailed review! This was left over from before it was decided to not include this, I'll remove it.

Commit:
Add "create deployment" sub-command

fabianofranz commented:
@asalkeld thanks!

LGTM, @smarterclayton please tag?
